### ISSUE TITLE
normalise layer

### ISSUE THREAD

**Issue by 1292765944** [happynear/NormFace#13]
I have a question about the normalise layer. In your paper, you use l2-normalise + scale layer before softmax. How about using batch-norm layers, which is more popular in modern DCNNs, e.g. ResNets. What's the difference of them?

Another question is the contrastive layer. In your C-contrastive loss and C-triplet loss, one of the input pairs is the input image, the others are the agent of the class. How do you implement the sampling and hard mining algorithms? Thank you!

Best!

**Comment by happynear** [comment]
Batch-norm is totally another thing. It normalize each **single** dimension of feature across all samples. It is not consistent with our goal which is described in Introduction. I use to try it, not because its normalization in scale, but its centralization because I also remove the mean value for each feature dimension during testing. However, finally I found that it had no influence on the performance.

There is no need to implement sampling and hard mining algorithm any more using C-contrastive and C-triplet loss. These two loss functions are created for classification, so just use them to replace the Softmax loss.

The codes for C-Contrastive loss and C-Triplet loss are in https://github.com/happynear/caffe-windows/blob/ms/src/caffe/layers/general_constrastive_loss_layer.cpp and https://github.com/happynear/caffe-windows/blob/ms/src/caffe/layers/general_triplet_loss_layer.cpp .

**Comment by 1292765944** [comment]
@happynear what does the positive_first mean? https://github.com/happynear/caffe-windows/blob/ms/src/caffe/layers/general_constrastive_loss_layer.cpp#L100
Best!

**Comment by happynear** [comment]
if the positive distance is above positive_upper_bound, optimize the positive distance only (set the gradients correspond to negative distances to be 0). 
There are lots of hyper-parameters in genereal_contrastive_loss_layer and general_triplet_loss_layer.   I added these hyper-parameters to explore the properties of these two loss functions just for research. For engineering purpose, you can ignore them and use the default settings.
